pkg/serviceaccount/metadata.go

```diff
@@ -0,0 +1,277 @@
+	w.Header().Set("Content-Type", "application/json")
+	// TODO(mtaufen): Set cache header appropriately
+
+	// TODO(mtaufen): trace the error handling here to understand how it works
+	enc := json.NewEncoder(w)
+	enc.SetIndent("", "\t")
+	// We can't stop writing the body and switch the header; but we can record when an error occurred.
+	err := enc.Encode(s.metadata)
+	if err != nil && s.errorHandler != nil {
+		s.errorHandler(ResponseError{
+			URL: *req.URL,
+			Err: err,
+		})
+	}
+}
+
+func getAlgs(keys *jose.JSONWebKeySet) []string {
+	var result []string
+	algs := map[string]bool{}
+	for _, k := range keys.Keys {
+		a := k.Algorithm
+		if !algs[a] {
+			result = append(result, a)
+		}
+		algs[a] = true
+	}
+	sort.Strings(result)
+	return result
+}
+
+func getJwks(in []interface{}) (*jose.JSONWebKeySet, error) {
+	// Decode keys into a JWKS.
+	var keys jose.JSONWebKeySet
+	var errs []error
+	for i, key := range in {
+		var pubkey *jose.JSONWebKey
+		var err error
+
+		switch k := key.(type) {
+		case interface {
+			Public() crypto.PublicKey
+		}:
+			// This is a private key. Get its public key
+			pubkey, err = jwkFromPubkey(k.Public())
+		case rsa.PublicKey:
+			pubkey, err = jwkFromPubkey(&k)
+		case *rsa.PublicKey:
+			pubkey, err = jwkFromPubkey(k)
+		case ecdsa.PublicKey:
+			pubkey, err = jwkFromPubkey(&k)
+		case *ecdsa.PublicKey:
+			pubkey, err = jwkFromPubkey(k)
+		default:
+			err = KeyExcludedError{
+				Key:    k,
+				Reason: "must be (*)rsa.PublicKey or (*)ecdsa.PublicKey",
+			}
+		}
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if !pubkey.Valid() {
+			errs = append(errs, KeyExcludedError{
+				Key:    pubkey,
+				Reason: fmt.Sprintf("configured key #%d not valid", i),
+			})
+			continue
+		}
+		keys.Keys = append(keys.Keys, *pubkey)
+	}
+	if len(errs) != 0 {
+		return &keys, fmt.Errorf("multiple errors: %v", errs)
+	}
+	return &keys, nil
+}
+
+func (s *issuerServer) serveKeys(w http.ResponseWriter, req *http.Request) {
+	// Per RFC7517 : https://tools.ietf.org/html/rfc7517#section-8.5.1
+	w.Header().Set("Content-Type", "application/jwk-set+json")
+	// TODO: Set cache header
+
+	enc := json.NewEncoder(w)
+	enc.SetIndent("", "\t")
+
+	err := enc.Encode(s.keys)
+
+	if err != nil {
+		s.handleError(ResponseError{
+			URL: *req.URL,
+			Err: err,
+		})
+	}
+}
+
+func jwkFromPubkey(k crypto.PublicKey) (*jose.JSONWebKey, error) {
+	alg, err := algorithmForPublicKey(k)
+	if err != nil {
+		return nil, KeyExcludedError{
+			Key:    k,
+			Reason: "must be *rsa.PublicKey, *ecdsa.PublicKey, or jose.OpaqueSigner",
+		}
+	}
+	return &jose.JSONWebKey{
+		Key:       k,
+		Algorithm: string(alg),
+		Use:       UseSigning,
+	}, nil
+}
+
+func serveError(w http.ResponseWriter, code int) {
+	http.Error(w, http.StatusText(code), code)
+}
+
+func (s *issuerServer) SetErrorHandler(f func(error)) {
+	s.errorHandler = f
+}
+
+func (s *issuerServer) handleError(err error) {
+	if err == nil {
```

Comment:
Why? This looks kind of weird. You are already checking this before calling the function. Just remove it from here.